File: pkg-r/tests/testthat/test-cache.R
test_that("cache mode checks existence against the bundled sample", {
  withr::local_envvar(BIOBOUNCER_CACHE_DIR = withr::local_tempdir())

  res <- check_id(
    c("MONDO:0005148", "MONDO:9999999", "mondo:5148"),
    source_db = "mondo",
    how = "cache",
    version = "sample"
  )
  expect_identical(res$valid, c(TRUE, FALSE, FALSE))
  expect_identical(res$normalized[1], "MONDO:0005148")
  expect_identical(res$suggestion[3], "MONDO:0005148")
  expect_identical(res$how, rep("cache", 3))
  expect_identical(res$version, rep("sample", 3))
})

test_that("a well-formed but absent suggestion is not offered in cache mode", {
  withr::local_envvar(BIOBOUNCER_CACHE_DIR = withr::local_tempdir())
  # mondo:9999999 suggests MONDO:9999999, which is not in the sample snapshot.
  res <- check_id(
    "mondo:9999999",
    source_db = "mondo",
    how = "cache",
    version = "sample"
  )
  expect_false(res$valid)
  expect_true(is.na(res$suggestion))
})

test_that("cache mode defaults to the latest installed snapshot", {
  withr::local_envvar(BIOBOUNCER_CACHE_DIR = withr::local_tempdir())
  # With no version, cache mode uses the latest installed snapshot instead of
  # forcing a magic version = "sample". The bundled sample is the only one here.
  res <- check_id("MONDO:0005148", source_db = "mondo", how = "cache")
  expect_true(res$valid)
  expect_identical(res$version, "sample")
})

test_that("cache default prefers the source's pinned default_version", {
  withr::local_envvar(BIOBOUNCER_CACHE_DIR = withr::local_tempdir())
  # hgnc pins default_version; the default resolves to it, not the sample.
  res <- check_id("TP53", source_db = "hgnc", how = "cache")
  expect_true(res$valid)
  expect_identical(res$version, "2026-07-07")
})

test_that("a defaulted cache check resolves to a newly bundled snapshot", {
  withr::local_envvar(BIOBOUNCER_CACHE_DIR = withr::local_tempdir())
  # doid now ships a bundled sample snapshot, so a defaulted cache check
  # resolves to it instead of erroring.
  res <- check_id("DOID:9352", source_db = "doid", how = "cache")
  expect_true(res$valid)
  expect_identical(res$version, "sample")
})

test_that("a missing snapshot is an actionable error", {
  withr::local_envvar(BIOBOUNCER_CACHE_DIR = withr::local_tempdir())
  expect_error(
    check_id(
      "MONDO:0005148",
      source_db = "mondo",
      how = "cache",
      version = "2099-01"
    ),
    class = "biobouncer_error_missing_snapshot"
  )
})

test_that("a traversal version is refused, not read", {
  withr::local_envvar(BIOBOUNCER_CACHE_DIR = withr::local_tempdir())
  for (bad in c("../../etc/passwd", "..\\..\\secret", "a/b", "sub/../x", "")) {
    expect_error(
      check_id(
        "MONDO:0005148",
        source_db = "mondo",
        how = "cache",
        version = bad
      ),
      class = "biobouncer_error_invalid_version"
    )
  }
  expect_error(
    biobouncer_pull("mondo", version = "../../evil"),
    class = "biobouncer_error_invalid_version"
  )
})

test_that("biobouncer_cache_dir honours the environment override", {
  d <- withr::local_tempdir()
  withr::local_envvar(BIOBOUNCER_CACHE_DIR = d)
  expect_identical(biobouncer_cache_dir(), d)
})

test_that("biobouncer_snapshots lists the bundled samples", {
  withr::local_envvar(BIOBOUNCER_CACHE_DIR = withr::local_tempdir())
  snaps <- biobouncer_snapshots()
  expect_s3_class(snaps, "tbl_df")
  expect_true(all(
    c("source", "version", "n_ids", "location") %in% names(snaps)
  ))
  mondo_sample <- snaps[snaps$source == "mondo" & snaps$version == "sample", ]
  expect_identical(nrow(mondo_sample), 1L)
  expect_gt(mondo_sample$n_ids, 0L)
  expect_identical(mondo_sample$location, "bundled")
})

test_that("biobouncer_snapshots lists the new obo samples", {
  withr::local_envvar(BIOBOUNCER_CACHE_DIR = withr::local_tempdir())
  snaps <- biobouncer_snapshots()
  bundled <- snaps$source[
    snaps$version == "sample" & snaps$location == "bundled"
  ]
  expect_true(all(
    c("bto", "cl", "doid", "hp", "mp", "pato", "so", "uberon") %in% bundled
  ))
})

File: pkg-py/tests/test_cache.py
"""Cache mode behavior against the bundled sample snapshot."""

import pytest

import biobouncer
from biobouncer import MissingSnapshotError


@pytest.fixture(autouse=True)
def _isolate_cache(tmp_path, monkeypatch):
    monkeypatch.setenv("BIOBOUNCER_CACHE_DIR", str(tmp_path))


def test_cache_checks_existence():
    res = biobouncer.check_id(
        ["MONDO:0005148", "MONDO:9999999", "mondo:5148"],
        source_db="mondo",
        how="cache",
        version="sample",
    )
    assert [r.valid for r in res] == [True, False, False]
    assert res[0].normalized == "MONDO:0005148"
    assert res[2].suggestion == "MONDO:0005148"
    assert all(r.how == "cache" for r in res)
    assert all(r.version == "sample" for r in res)


def test_absent_suggestion_is_not_offered():
    res = biobouncer.check_id(
        "mondo:9999999", source_db="mondo", how="cache", version="sample"
    )[0]
    assert res.valid is False
    assert res.suggestion is None


def test_cache_defaults_to_latest_installed_snapshot():
    # With no version, cache mode uses the latest installed snapshot instead of
    # forcing a magic version="sample". The bundled sample is the only one here.
    res = biobouncer.check_id("MONDO:0005148", source_db="mondo", how="cache")[0]
    assert res.valid is True
    assert res.version == "sample"


def test_cache_default_prefers_pinned_default_version():
    # hgnc pins default_version; the default resolves to it, not the sample.
    res = biobouncer.check_id("TP53", source_db="hgnc", how="cache")[0]
    assert res.valid is True
    assert res.version == "2026-07-07"


def test_cache_defaults_for_newly_snapshotted_obo_source():
    # doid now ships a bundled sample snapshot, so a defaulted cache check
    # resolves to it instead of erroring.
    res = biobouncer.check_id("DOID:9352", source_db="doid", how="cache")[0]
    assert res.valid is True
    assert res.version == "sample"


def test_missing_snapshot_errors():
    with pytest.raises(MissingSnapshotError):
        biobouncer.check_id(
            "MONDO:0005148", source_db="mondo", how="cache", version="2099-01"
        )


@pytest.mark.parametrize(
    "version", ["../../etc/passwd", "..\\..\\secret", "a/b", "sub/../x", ""]
)
def test_cache_rejects_traversal_version(version):
    # A version that could escape the snapshot directory is refused, not read.
    from biobouncer._cache import InvalidVersionError

    with pytest.raises(InvalidVersionError):
        biobouncer.check_id(
            "MONDO:0005148", source_db="mondo", how="cache", version=version
        )


def test_pull_rejects_traversal_version():
    from biobouncer._cache import InvalidVersionError

    with pytest.raises(InvalidVersionError):
        biobouncer.pull("mondo", version="../../evil")


def test_cache_dir_env_override(tmp_path, monkeypatch):
    monkeypatch.setenv("BIOBOUNCER_CACHE_DIR", str(tmp_path))
    assert biobouncer.cache_dir() == tmp_path


def test_snapshots_lists_bundled_sample():
    snaps = biobouncer.snapshots()
    mondo = [s for s in snaps if s["source"] == "mondo" and s["version"] == "sample"]
    assert len(mondo) == 1
    assert mondo[0]["location"] == "bundled"
    assert mondo[0]["n_ids"] > 0


def test_snapshots_include_the_new_obo_samples():
    snaps = biobouncer.snapshots()
    bundled = {
        s["source"]
        for s in snaps
        if s["version"] == "sample" and s["location"] == "bundled"
    }
    assert {"bto", "cl", "doid", "hp", "mp", "pato", "so", "uberon"} <= bundled
